File: app/src/main/java/com/ahmadabuhasan/appgithubuser/ui/UserDetailActivity.java
package com.ahmadabuhasan.appgithubuser.ui;

import static com.ahmadabuhasan.appgithubuser.db.DatabaseContract.FavoriteColumns.TABLE_NAME;

import android.content.Intent;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.ahmadabuhasan.appgithubuser.R;
import com.ahmadabuhasan.appgithubuser.adapter.ViewPagerAdapter;
import com.ahmadabuhasan.appgithubuser.databinding.ActivityUserDetailBinding;
import com.ahmadabuhasan.appgithubuser.db.DatabaseContract;
import com.ahmadabuhasan.appgithubuser.db.DatabaseHelper;
import com.ahmadabuhasan.appgithubuser.db.FavoriteHelper;
import com.ahmadabuhasan.appgithubuser.model.UserDetail;
import com.ahmadabuhasan.appgithubuser.viewmodel.UserViewModel;
import com.bumptech.glide.Glide;
import com.google.android.material.tabs.TabLayoutMediator;

import java.util.ArrayList;
import java.util.Objects;

import es.dmoral.toasty.Toasty;

public class UserDetailActivity extends AppCompatActivity {

    private ActivityUserDetailBinding binding;
    public static String DETAIL_USER = "DETAIL_USER";
    public static String dataUser;
    private UserViewModel userViewModel;
    private final int[] TAB_CLICK = new int[]{
            R.string.followers,
            R.string.following
    };

    ArrayList<UserDetail> userDetailArrayList = new ArrayList<>();
    private FavoriteHelper favoriteHelper;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_user_detail);
        setTitle(R.string.detail_user);

        Objects.requireNonNull(getSupportActionBar()).setHomeButtonEnabled(true);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        binding = ActivityUserDetailBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        favoriteHelper = FavoriteHelper.getFavoriteHelper(getApplicationContext());
        favoriteHelper.open();

        ViewPagerAdapter viewPagerAdapter = new ViewPagerAdapter(this);
        binding.viewPager.setAdapter(viewPagerAdapter);
        new TabLayoutMediator(binding.tabsLayout, binding.viewPager,
                (tab, position) -> tab.setText(getResources().getString(TAB_CLICK[position]))
        ).attach();

        dataUser = getIntent().getStringExtra(DETAIL_USER);
        showViewModel();
        userViewModel.isLoading().observe(this, this::showLoading);
    }

    private void showViewModel() {
        userViewModel = new ViewModelProvider(this, new ViewModelProvider.NewInstanceFactory()).get(UserViewModel.class);
        userViewModel.setUserDetail(dataUser);
        userViewModel.getUserDetail().observe(this, user -> {
            Glide.with(getApplicationContext())
                    .load(user.getAvatarUrl())
                    .into(binding.imgUser);
            binding.tvNameDetail.setText(user.getName());
            binding.tvUsernameDetail.setText(String.format("@%s", user.getUsername()));
            binding.tvCompanyDetail.setText(String.format("%s%s", getResources().getString(R.string.company), user.getCompany()));
            binding.tvLocationDetail.setText(String.format("%s%s", getResources().getString(R.string.location), user.getLocation()));
            binding.tvBlogDetail.setText(String.format("%s%s", getResources().getString(R.string.blog), user.getHtmlUrl()));
            binding.tvFollowers.setText(user.getFollowers());
            binding.tvRepository.setText(user.getRepository());
            binding.tvFollowing.setText(user.getFollowing());

            if (FavoriteExist(dataUser)) {
                binding.fabFavorite.setFavorite(true);
                binding.fabFavorite.setOnFavoriteChangeListener(
                        (buttonView, favorite) -> {
                            if (favorite) {
                                userDetailArrayList = favoriteHelper.getAllFavorite();
                                favoriteHelper.insertFavorite(user);
                                Toasty.success(getApplicationContext(), "Added Favorite", Toasty.LENGTH_SHORT).show();
                            } else {
                                userDetailArrayList = favoriteHelper.getAllFavorite();
                                favoriteHelper.deleteFavorite(dataUser);
                                Toasty.error(getApplicationContext(), "Deleted Favorite", Toasty.LENGTH_SHORT).show();
                            }
                        });
            } else {
                binding.fabFavorite.setOnFavoriteChangeListener(
                        (buttonView, favorite) -> {
                            if (favorite) {
                                userDetailArrayList = favoriteHelper.getAllFavorite();
                                favoriteHelper.insertFavorite(user);
                                Toasty.success(getApplicationContext(), "Added Favorite", Toasty.LENGTH_SHORT).show();
                            } else {
                                userDetailArrayList = favoriteHelper.getAllFavorite();
                                favoriteHelper.deleteFavorite(dataUser);
                                Toasty.error(getApplicationContext(), "Deleted Favorite", Toasty.LENGTH_SHORT).show();
                            }
                        });
            }
        });
    }

    public Boolean FavoriteExist(String user) {
        String select = DatabaseContract.FavoriteColumns.USERNAME + " =?";
        String[] Arg = {user};
        String limit = "1";
        favoriteHelper = new FavoriteHelper(this);
        favoriteHelper.open();
        DatabaseHelper dataBaseHelper = new DatabaseHelper(UserDetailActivity.this);
        SQLiteDatabase database = dataBaseHelper.getWritableDatabase();
        Cursor cursor = database.query(TABLE_NAME, null, select, Arg, null, null, null, limit);
        boolean exists;
        exists = (cursor.getCount() > 0);
        cursor.close();

        return exists;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.detail_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        } else if (item.getItemId() == R.id.share) {
            Intent i = new Intent(Intent.ACTION_VIEW);
            i.setData(Uri.parse("https://github.com/" + dataUser));
            startActivity(i);
        }
        return super.onOptionsItemSelected(item);
    }

    private void showLoading(Boolean isLoading) {
        if (isLoading) {
            binding.progressBar.setVisibility(View.VISIBLE);
        } else {
            binding.progressBar.setVisibility(View.GONE);
        }
    }
}

File: Submission-Kotlin/app/src/main/java/com/ahmadabuhasan/appgithubuser/ui/UserDetailActivity.kt
package com.ahmadabuhasan.appgithubuser.ui

import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.activity.viewModels
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.viewpager2.widget.ViewPager2
import com.ahmadabuhasan.appgithubuser.R
import com.ahmadabuhasan.appgithubuser.adapter.SectionsPagerAdapter
import com.ahmadabuhasan.appgithubuser.databinding.ActivityUserDetailBinding
import com.ahmadabuhasan.appgithubuser.db.DatabaseContract
import com.ahmadabuhasan.appgithubuser.db.DatabaseContract.FavoriteColumns.Companion.TABLE_NAME
import com.ahmadabuhasan.appgithubuser.db.DatabaseHelper
import com.ahmadabuhasan.appgithubuser.db.FavoriteHelper
import com.ahmadabuhasan.appgithubuser.model.ResponseDetailUser
import com.ahmadabuhasan.appgithubuser.viewmodel.MainViewModel
import com.bumptech.glide.Glide
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator

class UserDetailActivity : AppCompatActivity() {

    private val viewModel: MainViewModel by viewModels()
    private lateinit var binding: ActivityUserDetailBinding

    private var listFavorite = ArrayList<ResponseDetailUser>()
    private lateinit var helper: FavoriteHelper

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityUserDetailBinding.inflate(layoutInflater)
        setContentView(binding.root)

        supportActionBar?.title = resources.getString(R.string.detail_user)
        supportActionBar?.setHomeButtonEnabled(true)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val sectionsPagerAdapter = SectionsPagerAdapter(this)
        val viewPager: ViewPager2 = binding.viewPager
        viewPager.adapter = sectionsPagerAdapter
        val tabs: TabLayout = binding.tabs
        TabLayoutMediator(tabs, viewPager) { tab, position ->
            tab.text = resources.getString(TAB_TITLES[position])
        }.attach()

        username = intent.getStringExtra(EXTRA_USER).toString()
        showViewModel()
        viewModel.getIsLoading.observe(this, this::showLoading)

        helper = FavoriteHelper.getInstance(applicationContext)
        helper.open()
    }

    private fun showViewModel() {
        viewModel.detailUser(username)
        viewModel.getUserDetail.observe(this) { detailUser ->
            Glide.with(this)
                .load(detailUser.avatarUrl)
                .skipMemoryCache(true)
                .into(binding.imgAvatar)

            binding.tvName.text = detailUser.name
            binding.tvUsername.text = detailUser.login
            binding.tvCompany.text = detailUser.company
            binding.tvLocation.text = detailUser.location
            binding.tvRepositoryValue.text = detailUser.publicRepos
            binding.tvFollowersValue.text = detailUser.followers
            binding.tvFollowingValue.text = detailUser.following

            if (favoriteExist(username)) {
                binding.imageFavorite.isFavorite = true
                binding.imageFavorite.setOnFavoriteChangeListener { _, favorite ->
                    if (favorite) {
                        listFavorite = helper.queryAll()
                        helper.insert(detailUser)
                    } else {
                        listFavorite = helper.queryAll()
                        helper.delete(username)
                    }
                    helper.close()
                }
            } else {
                binding.imageFavorite.setOnFavoriteChangeListener { _, favorite ->
                    if (favorite) {
                        listFavorite = helper.queryAll()
                        helper.insert(detailUser)
                    } else {
                        listFavorite = helper.queryAll()
                        helper.delete(username)
                    }
                    helper.close()
                }
            }
        }
    }

    private fun favoriteExist(user: String): Boolean {
        val choose: String = DatabaseContract.FavoriteColumns.LOGIN + " =?"
        val chooseArg = arrayOf(user)
        val limit = "1"

        helper = FavoriteHelper(this)
        helper.open()

        val dataBaseHelper = DatabaseHelper(this@UserDetailActivity)
        val database: SQLiteDatabase = dataBaseHelper.writableDatabase
        val cursor: Cursor =
            database.query(TABLE_NAME, null, choose, chooseArg, null, null, null, limit)
        val exists: Boolean = cursor.count > 0
        cursor.close()

        database.close()
        return exists
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun showLoading(isLoading: Boolean) {
        if (isLoading) {
            binding.progressBar.visibility = View.VISIBLE
        } else {
            binding.progressBar.visibility = View.GONE
        }
    }

    companion object {
        const val EXTRA_USER = "extra_user"
        var username = String()

        @StringRes
        private val TAB_TITLES = intArrayOf(
            R.string.follower,
            R.string.following
        )
    }
}